reinforcement-learning/dqn.py

```diff
@@ -0,0 +1,317 @@
+
+      if len(self._rewards_list) >= self._num_rewards_to_average and (self._best_average is None or self._best_average < average):
+        self._best_average = average
+
+      if self._epsilon > self._final_epsilon:
+        self._epsilon -= self._epsilon_decay
+
+      print ('Episode', episode, 'Reward', self._total_reward, 'Average Reward', round(average, 2))
+      self._total_reward = 0.0
+      self._timesteps = 0
+
+    current_state = obs.reshape((1, len(obs)))
+
+    # Initialize the last state and action
+    if self._last_state is None:
+      self._last_state = current_state
+      q_values = self._neural_net.predict(session, self._last_state)
+      best_action = np.argmax(q_values)
+      self._last_action = np.zeros(self._action_space_dim)
+      self._last_action[best_action] = 1
+      return best_action, done
+
+    # Store the current step
+    new_step = [
+      self._last_state.copy(),
+      self._last_action.copy(),
+      reward,
+      current_state.copy(),
+      done
+    ]
+    self._replay_buffer.append(new_step)
+    self._last_state = current_state.copy()
+
+    # If the buffer is full, remove the oldest step
+    while len(self._replay_buffer) >= self._replay_memory_size:
+      self._replay_buffer.pop(0)
+
+    # Choose random actions during pure exploration
+    if episode < self._episodes_pure_exploration:
+      best_action = np.random.randint(0, self._action_space_dim)
+    else:
+      self._train(session)
+      # Use epsilon-greedy policy to choose best action
+      if np.random.random() > self._epsilon:
+        q_values = self._neural_net.predict(session, self._last_state)
+        best_action = np.argmax(q_values)
+      else:
+        best_action = np.random.randint(0, self._action_space_dim)
+
+    next_action = np.zeros(self._action_space_dim)
+    next_action[best_action] = 1
+    self._last_action = next_action
+
+    return best_action, done
+
+  def _train(self, session):
+
+    # Randomly sample from saved states
+    permutations = np.random.permutation(len(self._replay_buffer))[:self._minibatch_size]
+    minibatch = [self._replay_buffer[i] for i in permutations]
+
+    previous_states = np.concatenate([step[0] for step in minibatch])
+    actions = [step[1] for step in minibatch]
+    rewards = np.array([step[2] for step in minibatch]).astype('float')
+    current_states = np.concatenate([step[3] for step in minibatch])
+    done = np.array([step[4] for step in minibatch]).astype('bool')
+
+    # Current Q-values of the sampled states
+    q_values = self._neural_net.predict(session, current_states)
+
+    # Calculate target Q-values
+    target_q_values = rewards.copy()
+    target_q_values += ((1. - done) * self._discount_factor * q_values.max(axis=1))
+
+    # Update the Q function with new information
+    self._neural_net.fit(session, previous_states, actions, target_q_values)
+
+
+class QNetwork:
+
+  def __init__(self, layer_dims, assignments):
```

Comment:
why use `assignments` instead of `initial_weight_stddev, initial_bias_stddev, ...` as parameters?